pkg/controller/modify_status.go

```diff
@@ -0,0 +1,159 @@
+	pvcCondition := v1.PersistentVolumeClaimCondition{
+		Type:               v1.PersistentVolumeClaimVolumeModifyVolumeError,
+		Status:             v1.ConditionTrue,
+		LastTransitionTime: metav1.Now(),
+		Message:            "ModifyVolume failed with error. Waiting for retry.",
+	}
+
+	if err != nil {
+		pvcCondition.Message = "ModifyVolume failed with error: " + err.Error() + ". Waiting for retry."
+	}
+
+	newPVC.Status.Conditions = util.MergeModifyVolumeConditionsOfPVC(newPVC.Status.Conditions,
+		[]v1.PersistentVolumeClaimCondition{pvcCondition})
+
+	updatedPVC, err := ctrl.patchClaim(pvc, newPVC, false)
+	if err != nil {
+		return pvc, fmt.Errorf("mark PVC %q as controller expansion failed, errored with: %v", pvc.Name, err)
+	}
+	return updatedPVC, nil
+}
+
+// markControllerModifyVolumeStatus will mark ModifyVolumeStatus as completed in the PVC
+// and update CurrentVolumeAttributesClassName, clear the conditions
+func (ctrl *resizeController) markControllerModifyVolumeCompleted(pvc *v1.PersistentVolumeClaim, pv *v1.PersistentVolume) (*v1.PersistentVolumeClaim, *v1.PersistentVolume, error) {
+	// Update PVC
+	newPVC := pvc.DeepCopy()
+	// Update ModifyVolumeStatus to completed
+	var controllerModifyVolumeCompleted v1.PersistentVolumeClaimModifyVolumeStatus
+	if newPVC.Status.ModifyVolumeStatus == nil {
+		newPVC.Status.ModifyVolumeStatus = &v1.ModifyVolumeStatus{}
+	}
+	newPVC.Status.ModifyVolumeStatus.Status = controllerModifyVolumeCompleted
+
+	// Update CurrentVolumeAttributesClassName
+	newPVC.Status.CurrentVolumeAttributesClassName = &newPVC.Status.ModifyVolumeStatus.TargetVolumeAttributesClassName
+
+	// Clear all the conditions
+	var err error
+	newPVC.Status.Conditions, err = clearModifyVolumeConditions(newPVC.Status.Conditions)
+
+	updatedPVC, err := ctrl.patchClaim(pvc, newPVC, false)
+
+	if err != nil {
+		return pvc, pv, fmt.Errorf("mark PVC %q as ModifyVolumeCompleted failed, errored with: %v", pvc.Name, err)
+	}
+
+	// Update PV
+	newPV := pv.DeepCopy()
+	newPV.Spec.VolumeAttributesClassName = &newPVC.Status.ModifyVolumeStatus.TargetVolumeAttributesClassName
+
+	updatedPV, err := ctrl.patchPersistentVolume(pv, newPV)
+	if err != nil {
+		return pvc, pv, fmt.Errorf("update pv.Spec.VolumeAttributesClassName for PVC %q failed, errored with: %v", pvc.Name, err)
+	}
+
+	return updatedPVC, updatedPV, nil
+}
+
+// markControllerModifyVolumeStatus clears all the conditions related to modify volume and only
+// leave other condition types
+func clearModifyVolumeConditions(conditions []v1.PersistentVolumeClaimCondition) ([]v1.PersistentVolumeClaimCondition, error) {
+	cleanedConditions := []v1.PersistentVolumeClaimCondition{}
+	for _, value := range conditions {
+		// Only keep conditions that are not related to modify volume
+		if value.Type != v1.PersistentVolumeClaimVolumeModifyVolumeError && value.Type != v1.PersistentVolumeClaimVolumeModifyingVolume {
+			cleanedConditions = append(cleanedConditions, value)
```

Comment:
may be we can use `knownConditions` here?

Reply:
Updated